packages/legend-query-builder/src/components/QueryLoaderState.ts

```diff
@@ -0,0 +1,290 @@
+ * limitations under the License.
+ */
+
+import {
+  DEFAULT_TYPEAHEAD_SEARCH_MINIMUM_SEARCH_LENGTH,
+  DEFAULT_TYPEAHEAD_SEARCH_LIMIT,
+  type GenericLegendApplicationStore,
+} from '@finos/legend-application';
+import {
+  type LightQuery,
+  QuerySearchSpecification,
+  type AbstractPureGraphManager,
+  type QueryInfo,
+  type QueryTaggedValue,
+  type QueryProjectCoordinates,
+  V1_PureGraphManager,
+  type Query,
+} from '@finos/legend-graph';
+import {
+  ActionState,
+  type GeneratorFn,
+  assertErrorThrown,
+  guaranteeType,
+  deleteEntry,
+} from '@finos/legend-shared';
+import { makeObservable, observable, action, flow, flowResult } from 'mobx';
+import { QUERY_BUILDER_SETTING_KEY } from '../application/QueryBuilderSetting.js';
+
+type QueryPersistInfo = {
+  id: string;
+  owner: string | undefined;
+};
+
+export class QueryLoaderState {
+  readonly applicationStore: GenericLegendApplicationStore;
+  loadQueriesState = ActionState.create();
+  queries: LightQuery[] = [];
+  isQueryLoaderOpen = false;
+  searchText = '';
+  showCurrentUserQueriesOnly = false;
+  showPreviewViewer = false;
+  querypreviewContent?: QueryInfo;
+  searchSpecificationFilters?: QueryTaggedValue[];
+  searchSpecificationProjectCoordinates?: QueryProjectCoordinates[];
+  currentUserId: string | undefined;
+  persistedQueries: QueryPersistInfo[] = [];
+
+  constructor(applicationStore: GenericLegendApplicationStore) {
+    makeObservable(this, {
+      isQueryLoaderOpen: observable,
+      querypreviewContent: observable,
+      queries: observable,
+      showCurrentUserQueriesOnly: observable,
+      showPreviewViewer: observable,
+      searchText: observable,
+      setSearchText: action,
+      setIsQueryLoaderOpen: action,
+      setQueries: action,
+      deleteQuery: action,
+      setShowCurrentUserQueriesOnly: action,
+      setShowPreviewViewer: action,
+      persistQueryIds: action,
+      loadQueries: flow,
+      getPreviewQueryContent: flow,
+      initializeWithCustomQueries: flow,
+      initialize: flow,
+    });
+    this.applicationStore = applicationStore;
+    const persistedQueries =
+      this.applicationStore.settingService.getObjectValue(
+        QUERY_BUILDER_SETTING_KEY.RECENTLY_VIEWED_QUERIES,
+      );
+    this.persistedQueries = persistedQueries
+      ? (persistedQueries as QueryPersistInfo[])
+      : [];
+  }
+
+  setSearchText(val: string): void {
+    this.searchText = val;
+  }
+
+  setIsQueryLoaderOpen(val: boolean): void {
+    this.isQueryLoaderOpen = val;
+  }
+
+  setQueries(val: LightQuery[]): void {
+    this.queries = val;
+  }
+
+  setShowPreviewViewer(val: boolean): void {
+    this.showPreviewViewer = val;
+  }
+
+  setShowCurrentUserQueriesOnly(val: boolean): void {
+    this.showCurrentUserQueriesOnly = val;
+  }
+
+  deleteQuery(query: LightQuery): void {
+    deleteEntry(this.queries, query);
+  }
+
+  persistQueryIds(val: LightQuery | Query): void {
```

Comment:
Could we try to make this method a bit more easier to read/understand the logic. 
Even just addingcomments. 
we persist the query only if its present or if its not present the length of the queries is shorter than DEFAULT_TYPEAHEAD_SEARCH_LIMIT ?